# -*- snakemake -*-
#
# Manual installation required. No setup exists so full path to
# installation directory must be supplied. Download from
# https://sites.google.com/site/danposdoc/download and extract archive
#
# Requirements:
# R version 2.9.1.
# Python 2.7, rpy2, numpy 1.5.0.
# Samtools 0.1.7 (only when input data is in sam or bam format)
#
# Memory ~ (genome_size/step_size) x ( replicate_count + max(2,
# comparison_count) ) x 8 bits.
#
import pysam
from snakemakelib.config import update_snakemake_config
from snakemakelib.stat import is_installed

include: "../settings.rules"

config_default = {
    'bio.ngs.chromatin.danpos' : {
        'home' : "",
        'cmd' : 'danpos.py',
        'bins' : [(0,100), (180,247), (315,473), (558,615)],
        'dpos' : {
            'options' : "",
        },
        'dpeak' : {
            'options' : "",
        },
        'dregion' : {
            'options' : "",
        },
        'dtriple' : {
            'options' : "",
        },
        'profile' : {
            'options' : "",
        },
        'wig' : {
            'options' : "",
        },
        'wig2wiq' : {
            'options' : "",
        },
        'stat' : {
            'options' : "",
        },
        'selector' : {
            'options' : "",
        },
        'valuesAtRanks' : {
            'options' : "",
        },
    },
}

config = update_snakemake_config(config, config_default)

dp_cfg = config['bio.ngs.chromatin.danpos']
python2 = config['comp.settings']['python2']['activate_cmd']

rule danpos_bin_reads_by_nucleosome:
    """Split reads into bins base on the parameter 'bins'"""
    input: bam="{prefix}.bam"
    output: bed=[temporary("{{prefix}}.{nucleosome}.bed".format(nucleosome=x)) for x in range(len(dp_cfg['bins']))]
    run:
        print (output.bed)
        # samfile = pysam.AlignmentFile(input.bam, "rb")
        # for s in samfile:
        #     print (dir(s))
        #     break
